### ISSUE TITLE
Visualizing images while training with TensorBoard

### ISSUE THREAD

**Issue by blakenguyen97** [google/automl#396]
How can I visualize the images after evaluating with Tensorboard like we usually see when training with Object detection API? Thank you

![image](https://user-images.githubusercontent.com/52020800/81635827-72037800-943c-11ea-8507-3f0279e8d026.png)

**Comment by mingxingtan** [comment]
Good idea. We should add that. Do you want to give it a try? 
I can also add that later this week.

**Comment by blakenguyen97** [comment]
> Good idea. We should add that. Do you want to give it a try?
> I can also add that later this week.

Ok I will give it a try. Thank you @mingxingtan :D

**Comment by jdvylder** [comment]
> > Good idea. We should add that. Do you want to give it a try?
> > I can also add that later this week.
> 
> Ok I will give it a try. Thank you @mingxingtan :D

This would be really useful. Currently I don't get the expected results, and based on some other issues reporting low Ap values I'm not the only one, which makes me wonder if my data pipeline is correct. This kind of visual feedback would greatly facilitate debugging.

**Comment by Samjith888** [comment]
> Good idea. We should add that. Do you want to give it a try?
> I can also add that later this week.

Any update on this feature ?

**Comment by mingxingtan** [comment]
Would be great if anyone can help add this feature. Thanks!

**Comment by kahkeng** [comment]
I see some code that seems to already exist in the repo that could be helpful for this purpose, but nothing seems to be calling it right now. E.g. in `efficientdet/visualize/vis_utils.py`, we have the function `draw_side_by_side_evaluation_image` being invoked by `VisualizeSingleFrameDetections`. This function takes `eval_dict` which is documented as the following:
```
    eval_dict: The evaluation dictionary returned by
      eval_util.result_dict_for_batched_example() or
      eval_util.result_dict_for_single_example().
```

However, I can't find the `eval_util` module in this repo. If we had some pointers for what this module does, or if this file could be checked in to this repo, it would be easier to hook it up with existing visualization code.

**Comment by wihanbooyse** [comment]
Hi, first of all thanks for the great work. Efficientdet really is a gamechanger. 

I have a workaround implemented for this that involves using eval hooks. Unfortunately I cannot get it working on a TPU on aiplatform as theres, issues with the XLA compilation of Non_max_suppressionv5. I've also experimented with using the Host call, but unfortunately setting image_summary_steps to 500 reduces training speed by a factor of 4-5. At which point it's a waste of TPU power.

If the ability to perform Eval visualisations as a separate process when running TPU training, and conventional train and eval when running GPU (Only tested on single GPU) would be useful to the community and is an acceptable first step to resolution of this this issue. Then i can put some effort into building some tests and creating a pull request for this?

**Comment by mingxingtan** [comment]
Hi @wihanbooyse Thank you for the info. It would be great if you can prepare  a PR for the first step.
I and other people can help solve the other issues (like the Non_max_suppressionv5 issue)

**Comment by fsx950223** [comment]
It's easy to implement it by async subprocess

**Comment by kartik4949** [comment]
@mingxingtan @blakenguyen97 solved this in recent PR #865 , please use it for visualizing train or eval records before training with proper config file.
